cpp/include/raft/spatial/knn/detail/ann_ivf_flat.cuh

```diff
@@ -0,0 +1,709 @@
+                    trainset_labels.data(),
+                    metric_type_,
+                    true,
+                    mc_trainset_ccenters_tmp.data(),
+                    mc_trainset_csizes_tmp.data(),
+                    true,
+                    stream_,
+                    &kmeans_mem_res);
+  }  // end for (int iter = 0; iter < 2; iter++)
+
+  RAFT_LOG_DEBUG("(%s) Final fitting.", __func__);
+
+  kmeans::predict(handle_,
+                  cluster_centers,
+                  n_lists_,
+                  dim_,
+                  dataset,
+                  n_rows_,
+                  labels,
+                  metric_type_,
+                  true,
+                  mc_trainset_ccenters_tmp.data(),
+                  mc_trainset_csizes_tmp.data(),
+                  true,
+                  stream_,
+                  &kmeans_mem_res);
+
+  kmeans::predict(handle_,
+                  cluster_centers,
+                  n_lists_,
+                  dim_,
+                  dataset,
+                  n_rows_,
+                  labels,
+                  metric_type_,
+                  true,
+                  mc_trainset_ccenters_tmp.data(),
+                  mc_trainset_csizes_tmp.data(),
+                  false,
+                  stream_,
+                  &kmeans_mem_res);
+}
+
+template <typename T>
+__global__ void build_index_kernel(const uint32_t* labels,
+                                   const uint32_t* list_offsets,
+                                   const T* dataset,
+                                   T* list_data,
+                                   uint32_t* list_index,
+                                   uint32_t* list_lengths,
+                                   uint32_t n_rows,
+                                   uint32_t dim,
+                                   uint32_t veclen)
+{
+  const int i = blockDim.x * blockIdx.x + threadIdx.x;
+  if (i >= n_rows) { return; }
+
+  auto list_id     = labels[i];
+  auto inlist_id   = atomicAdd(list_lengths + list_id, 1);
+  auto list_offset = list_offsets[list_id];
+
+  // Record the source vector id in the index
+  list_index[list_offset + inlist_id] = i;
+
+  // The data is written in interleaved groups of `WarpSize` vectors
+  using interleaved_group = Pow2<WarpSize>;
+  auto group_offset       = interleaved_group::roundDown(inlist_id);
+  auto ingroup_id         = interleaved_group::mod(inlist_id) * veclen;
+
+  // Point to the location of the interleaved group of vectors
+  list_data += (list_offset + group_offset) * dim;
+
+  // Point to the source vector
+  dataset += i * dim;
+
+  // Interleave dimensions of the source vector while recording it.
+  for (uint32_t l = 0; l < dim; l += veclen) {
+    for (uint32_t j = 0; j < veclen; j++) {
+      list_data[l * WarpSize + ingroup_id + j] = dataset[l + j];
+    }
+  }
+}
+
+template <typename T>
+void cuivflHandle<T>::cuivflBuildIndex(const T* dataset,
+                                       T* trainset,
```

Comment:
Do we need both the dataset (I'm assuming this would be for query?) and the trainset here? Why not just use the trainset?

Reply:
In the current implementation, `trainset` is a smaller subset of the `dataset` that is used to train k-means. Supposedly, reducing the size of the dataset makes it faster without sacrificing the quality too much.
After that, we use the whole dataset to build the index, which is just grouping the points in clusters (we need all points there, since they are what kNN returns).

Reply:
I see- I think `approx_knn_cuivfl_ivfflat_build_index` is probably what we want to invoke from FAISS. This method is lower level.